File: solutions/typescript/bowling/1/bowling.ts

export class Bowling {
  private static readonly LAST_FRAME = 9;
  private static readonly MAX_PINS = 10;
  private rolls: number[] = [];
  private currentFrame: number = 0;
  private frameRolls: number[] = [];
  private isGameOver: boolean = false;

  public roll(pins: unknown): void {
    if (this.isGameOver) throw new Error('Cannot roll after game is over');
    if (typeof pins !== 'number' || !Number.isInteger(pins)) throw new Error('Invalid roll');
    if (pins < 0) throw new Error('Negative roll is invalid');
    if (pins > Bowling.MAX_PINS) throw new Error('Pin count exceeds pins on the lane');

    // Validate bonus rolls in the 10th frame
    if (this.currentFrame === Bowling.LAST_FRAME) {
      this.validateBonusRolls(pins);
    }

    // Validate frame pin count except for strikes and 10th frame
    if (this.currentFrame < Bowling.LAST_FRAME && this.frameRolls.length === 1 && this.frameRolls[0] + pins > Bowling.MAX_PINS) {
      throw new Error('Pin count exceeds pins on the lane');
    }

    this.rolls.push(pins);
    this.frameRolls.push(pins);

    // Frame progression
    if (this.currentFrame < Bowling.LAST_FRAME) {
      if (pins === Bowling.MAX_PINS || this.frameRolls.length === 2) {
        this.currentFrame++;
        this.frameRolls = [];
      }
      return;
    }
    // 10th frame logic
    if (this.frameRolls.length === 2) {
      if (this.frameRolls[0] === Bowling.MAX_PINS || this.frameRolls[0] + this.frameRolls[1] === Bowling.MAX_PINS) {
        // Allow up to 3 rolls in 10th frame
        return;
      }
      this.isGameOver = true;
    } else if (this.frameRolls.length === 3) {
      this.isGameOver = true;
    }
  }

  public score(): number {
    if (this.rolls.length === 0) throw new Error('Score cannot be taken until the end of the game');
    if (!this.isGameComplete()) throw new Error('Score cannot be taken until the end of the game');
    let score = 0;
    let rollIndex = 0;
    for (let frame = 0; frame <= Bowling.LAST_FRAME; frame++) {
      if (this.rolls[rollIndex] === Bowling.MAX_PINS) {
        score += Bowling.MAX_PINS + this.strikeBonus(rollIndex);
        rollIndex += 1;
      } else if (this.rolls[rollIndex] + this.rolls[rollIndex + 1] === Bowling.MAX_PINS) {
        score += Bowling.MAX_PINS + this.spareBonus(rollIndex);
        rollIndex += 2;
      } else {
        score += this.rolls[rollIndex] + this.rolls[rollIndex + 1];
        rollIndex += 2;
      }
    }
    return score;
  }

  private strikeBonus(rollIndex: number): number {
    return (this.rolls[rollIndex + 1] ?? 0) + (this.rolls[rollIndex + 2] ?? 0);
  }

  private spareBonus(rollIndex: number): number {
    return this.rolls[rollIndex + 2] ?? 0;
  }

  private isGameComplete(): boolean {
    // 10th frame logic
    let frame = 0;
    let rollIndex = 0;
    while (frame < 10 && rollIndex < this.rolls.length) {
      if (this.rolls[rollIndex] === 10) {
        // Strike
        rollIndex += 1;
      } else {
        rollIndex += 2;
      }
      frame++;
    }
    // If not enough rolls for 10 frames
    if (frame < 10) return false;
    // 10th frame bonus logic
    const tenthStart = this.getFrameStart(9);
    const tenth = this.rolls.slice(tenthStart);
    if (tenth.length < 2) return false;
    if (tenth[0] === 10) {
      if (tenth.length < 3) return false;
    } else if (tenth[0] + (tenth[1] ?? 0) === 10) {
      if (tenth.length < 3) return false;
    }
    return true;
  }

  private getFrameStart(frame: number): number {
    let rollIndex = 0;
    let currentFrame = 0;
    while (currentFrame < frame && rollIndex < this.rolls.length) {
      if (this.rolls[rollIndex] === 10) {
        rollIndex += 1;
      } else {
        rollIndex += 2;
      }
      currentFrame++;
    }
    return rollIndex;
  }

  private validateBonusRolls(pins: number): void {
    const tenthFrame = this.rolls.slice(this.getFrameStart(9));
    if (tenthFrame.length === 1) {
      // First bonus roll after strike or spare
      if (tenthFrame[0] === 10 && pins > 10) {
        throw new Error('Pin count exceeds pins on the lane');
      }
    } else if (tenthFrame.length === 2) {
      if (tenthFrame[0] === 10) {
        // Two bonus rolls after strike
        if (pins > 10) {
          throw new Error('Pin count exceeds pins on the lane');
        }
        if (tenthFrame[1] < 10) {
          // If first bonus is not a strike, second cannot be a strike
          if (pins === 10) {
            throw new Error('Pin count exceeds pins on the lane');
          }
          // Their sum cannot exceed 10
          if (tenthFrame[1] + pins > 10) {
            throw new Error('Pin count exceeds pins on the lane');
          }
        }
      } else if (tenthFrame[0] + tenthFrame[1] === Bowling.MAX_PINS) {
        // Bonus roll after spare
        if (pins > 10) {
          throw new Error('Pin count exceeds pins on the lane');
        }
      }
    }
  }
}

File: solutions/elixir/bowling/1/lib/bowling.ex
defmodule Bowling do
  @doc """
    Creates a new game of bowling that can be used to store the results of
    the game
  """

  @spec start() :: any
  def start do
    %{complete: false, curr_frame: 0, frames: List.duplicate([], 10)}
  end

  @doc """
    Records the number of pins knocked down on a single roll. Returns `any`
    unless there is something wrong with the given number of pins, in which
    case it returns a helpful error tuple.
  """

  @spec roll(any, integer) :: {:ok, any} | {:error, String.t()}
  def roll(%{complete: complete}, _) when complete, do: {:error, "Cannot roll after game is over"}
  def roll(_, roll) when roll < 0, do: {:error, "Negative roll is invalid"}
  def roll(_, roll) when roll > 10, do: {:error, "Pin count exceeds pins on the lane"}

  def roll(game, roll) when game.curr_frame == 9 do
    # allow for up to 2 extra balls
    frame = case Enum.at(game.frames, game.curr_frame) do
      [] -> [roll]
      [first] -> [first, roll]
      [first, second] when first + second >= 10 ->
        [first, second, roll]
    end

    if (! validate_last_frame(frame)) do
      {:error, "Pin count exceeds pins on the lane"}
    else
      {:ok, update_game_state(game, frame)}
    end
  end

  def roll(game, roll) do
    frame = case Enum.at(game.frames, game.curr_frame) do
      [] -> [roll]
      [first] -> [first, roll]
    end
    if Enum.sum(frame) > 10 do
      {:error, "Pin count exceeds pins on the lane"}
    else
      {:ok, update_game_state(game, frame)}
    end
  end

  @doc """
    Returns the score of a given game of bowling if the game is complete.
    If the game isn't complete, it returns a helpful error tuple.
  """

  @spec score(any) :: {:ok, integer} | {:error, String.t()}
  def score(game) do
    if game.complete do
      score =
        game.frames
        |> Enum.with_index()
        |> Enum.reduce(0, fn {frame, i}, acc ->
          case frame do
            [10] ->
              acc + 10 + score_strike_bonus(game.frames, i + 1)
            [first, second] when first + second == 10 ->
              acc + first + second + Enum.at(Enum.at(game.frames, i + 1), 0)
            [first, second] -> acc + first + second
            [first, second, third] -> acc + first + second + third
          end
        end)
      {:ok, score}
    else
      {:error, "Score cannot be taken until the end of the game"}
    end
  end

  defp score_strike_bonus(frames, i) do
    bonus1 = Enum.at(Enum.at(frames, i), 0)
    bonus2 = if bonus1 == 10 and i < 9 do
      Enum.at(Enum.at(frames, i + 1), 0)
    else
      Enum.at(Enum.at(frames, i), 1)
    end
    bonus1 + bonus2
  end

  defp update_game_state(game, frame) do
    frames = List.update_at(game.frames, game.curr_frame, fn _ -> frame end)
    curr_frame = case frame do
      [10] when game.curr_frame == 9 -> game.curr_frame
      [10] -> game.curr_frame + 1
      [roll] when roll < 10 -> game.curr_frame
      [r1, r2] when game.curr_frame == 9 and r1 + r2 >= 10 -> game.curr_frame
      [_, _] -> game.curr_frame + 1
      [_, _, _] -> game.curr_frame + 1
    end
    complete = case curr_frame do
      10 -> true
      _ -> false
    end
    %{game | complete: complete, curr_frame: curr_frame, frames: frames}
  end

  defp validate_last_frame(frame) do
    case frame do
      [_] -> true
      [10, _] -> true
      [10, 10, _] -> true
      [10, first, second] when first + second <= 10 -> true
      [10, first, second] when first + second > 10 -> false
      [first, second, third] when first + second == 10 and third <= 10 -> true
      [first, second] -> (first + second) <= 10
    end
  end
end
